### ISSUE TITLE
[UX] Clean up menu block descriptions

### ISSUE THREAD

**Issue by jenlampton** [backdrop/backdrop-issues#2324]
Every menu block description is "The hierarchical list of links for the ___ menu" but now that we have an option for these to display differently, we should remove "hierarchical" from that description. How about just "A list of links in the ___ menu"?

---
~~PR by @klonos: https://github.com/backdrop/backdrop/pull/1636~~
~~PR by @BWPanda: https://github.com/backdrop/backdrop/pull/2697~~
PR by @BWPanda/@klonos: https://github.com/backdrop/backdrop/pull/2715

**Comment by klonos** [comment]
Good catch @jenlampton 👍 

> audience - novice

Me, me again 😄 ...

**Comment by klonos** [comment]
I went "fancy" with my php again 😜 ...it might seem crazy that I first remove the word "menu" from the menu title only to add it again later in the menu description, but that has the benefit of adding the word "menu" to menus that don't have it ("Primary navigation" becomes "Primary navigation menu") while not adding it twice for menus that do have it (Administration menu, Account menu). "And what does that gain us?" you ask...

Before:

![backdrop-add_block_dialog-before_pr1636](https://cloud.githubusercontent.com/assets/2423362/20002578/a793d8c2-a2d5-11e6-89d1-9eaaf2ec9793.png)

After:

![backdrop-add_block_dialog-after_pr1636](https://cloud.githubusercontent.com/assets/2423362/20002586/aee20ca2-a2d5-11e6-9247-32d149dfb462.png)

Notice how before, if one searched for the term "menu" in the "Add block" dialog, "Primary navigation" would not come up?

I've also changed the `@title` in the `t()` functions to `%title` to render the menu titles wrapped in `<em>` tags.

**Comment by jenlampton** [comment]
Looking good :)

- -
Please excuse brevity and typos. Message typed with my thumbs from an Open
Source OS.

**Comment by klonos** [comment]
Thanx for taking a look @jenlampton.

Quick question: the description text "**A** list of links for the `%menu` menu." in `system.module` renders the descriptions for the menu blocks in the "Manage blocks" page as well as in the "Add block" dialog. I don't know where the "**The** list of links for the `%title` menu." in `menu.module` is used though?

I thought that it would be great if there was a "Rendered as: `[menu-type]`" text added to the blocks in the "Manage blocks" page. That way people could get an overview of whether the menu block is set to drop-down, top-level only or tree without having to either open the block configuration dialog or visit the front-end to actually see how it is rendered for the specific layout they are editing (similar to the visibility conditions summary), but a) I thought "scope creep" and b) my coding skills were not strong enough to figure out how to pull that information and and where to add the code that renders it. It would have to be in the block description on the "Manage blocks" page only and not on the descriptions that are shown in the "Add block" dialog. Perhaps I'll keep exploring that in a separate issue.

**Comment by jenlampton** [comment]
The text you see in the "Add block" dialog is generated elsewhere than the text you see on the "Manage blocks" page -- even though it may be the same text.  Let's start with the change to `menu_block_info()` and close this issue.  But let's open a new one for the change on the manage blocks page, as that's a good suggestion too :)

**Comment by klonos** [comment]
> The text you see in the "Add block" dialog is generated elsewhere than the text you see on the "Manage blocks" page

Hmm, but why then if I change the description text in `menu_block_info()` in https://github.com/backdrop/backdrop/blob/1.x/core/modules/menu/menu.module#L455 I see no change in the "Add block" dialog and no change in the "Manage blocks" page either?

On the other hand, if I change the description text in `system_block_info()` in https://github.com/backdrop/backdrop/blob/1.x/core/modules/system/system.module#L2352 then I see the change reflected in *both* the "Add block" dialog *and* the "Manage blocks" page.

**Comment by klonos** [comment]
...that's why I asked "I don't know where the "**The** list of links for the `%title` menu." in `menu.module` is used though?" in the first place. I don't think that the description text there is used anywhere.

**Comment by klonos** [comment]
...in fact, in https://github.com/backdrop/backdrop/blob/1.x/core/modules/layout/includes/block.class.inc#L146 the `getAdminPreview()` function seems to be calling `getBlockInfo()`, getting its values in a `$info` array and then printing `$info['description']` as the admin preview. If I'm not misunderstanding all this, that make the admin preview text (in the "Manage blocks" page) the same as the block description (which is what is shown in the "Add block" dialog). Right?

**Comment by jenlampton** [comment]
> why then ...

Ah, Right. System module provides the default menus (Administration, User menu, Primary Navigation). Menu module provides user-created menus. If you create some new menus (I think) you'll see that those are controlled by the text in menu.module. I personally think this is silly, and have created an issue (somewhere) to move the menu stuff out of system module and back into menu module. (In Drupal 8 Menu module was renamed to "menu UI", a name that more closely matches what it does.)

>  that makes the admin preview text (in the "Manage blocks" page) the same as the block description

Yes, the text is the same. It even comes form the same place in the code. My point was that to change what's on the manage blocks page we don't want to change what's defined in hook_block_info(), because that would also change what's in the block list. Instead we should change what's in the block class, and add the additional settings/summary there. So, separate issue for that, maybe.

**Comment by klonos** [comment]
> Ah, Right. System module provides the default menus (Administration, User menu, Primary Navigation). Menu module provides user-created menus. If you create some new menus (I think) you'll see that those are controlled by the text in menu.module.

You are absolutely right. Just for testing purposes, I've added the word "system" in the system.module description and the word "custom" in the menu.module description and here's the result:

![backdrop-issue2324-system_vs_custom_menus](https://cloud.githubusercontent.com/assets/2423362/20024079/38fa5176-a338-11e6-802d-e416a9657269.png)

...did not thing of that but it makes perfect sense now. Thanx for the pointer @jenlampton 👍 

> I personally think this is silly, and have created an issue (somewhere) to move the menu stuff out of system module and back into menu module. (In Drupal 8 Menu module was renamed to "menu UI", a name that more closely matches what it does.)

Yeah, having to change any common properties (descriptions for example) separately for system vs. custom menus makes no sense to me DX-wise unless there is a very particular reason for it.

> My point was that to change what's on the manage blocks page we don't want to change what's defined in hook_block_info(), because that would also change what's in the block list.

I understand, but if I only change the description text in the menu.module, then system menu blocks will still have the silly "*hierarchical* list of links". So it needs to be changed in both places. Right?

**Comment by klonos** [comment]
> I personally think this is silly, and have created an issue (somewhere) to move the menu stuff out of system module and back into menu module. (In Drupal 8 Menu module was renamed to "menu UI", a name that more closely matches what it does.)

I think that also makes sense. @docwilmot has created the very related #1802 ...and by the way, I've found your issue (and it's a 3-digit one - one of the low): #198

**Comment by jenlampton** [comment]
> I understand, but if I only change the description text in the menu.module, then system menu blocks will still have the silly "hierarchical list of links". So it needs to be changed in both places. Right?

Right! :)

**Comment by klonos** [comment]
OK, I will close the current PR and file a new one. Reason is that my current approach has a few flaws:

- If the menu does not include the word "menu" in its title, then it will be added to the description. Example:
  "A list of links for the Primary navigation" would become "A list of links for the *Primary navigation* menu" <- OK, the menu title in this sentence is in italics and the word "menu" is added at the end of the sentence.
- If the menu has the word "menu" already included in its title, then I would trim it from the menu title and let it be added only once in the description. Example:
  "A list of links for the Administration menu" would become "A list of links for the *Administration* menu" <- OK, the word "menu" is not added twice, BUT it is not in italics. It should be trimmed from the block description text instead of the menu title.
- I did not account for the cases when a custom block has the word "Menu" (capital "M"). I should have used `str_ireplace` instead of `str_replace`.
- If a custom menu title is in the form of "myCustomMenu", then the word menu would still get added in the description, resulting in "A list of links for the *myCustomMenu* menu". Not sure about this one. I lean towards leaving that as is.

Anyways, I think I have a better idea of how to handle this more intuitively and I'll file another PR. After that, I'll try to solve in the same manner the similar issue of text duplicating the word "block" as in "The Hero block block".

**Comment by quicksketch** [comment]
Moving this back into needs work then. Thanks for your work on this @klonos!

**Comment by BWPanda** [comment]
Found this issue after struggling to find the "Primary Navigation" block when searching for 'menu'. I support at least adding the word 'menu' to each menu block so that searches display as expected. Haven't read through the rest of the issues mentioned here though...

**Comment by BWPanda** [comment]
I'm revitalising this issue with a fresh [PR](https://github.com/backdrop/backdrop/pull/2697) based on @klonos' work and notes above (hope you don't mind). I think I've addressed all the issues you had with your initial attempt, but feedback welcome how to improve it more if necessary.

Regarding the decision to have almost duplicated descriptions and choose between them (rather than simply adding 'menu' onto the end as necessary), this was the result of a discussion in Gitter about the correct way to do this (allowing for the full-stop, and making sure the word 'menu' on the end was translatable as well). If you'd like to read the conversation, search Gitter's history for 'issue-2324' then scroll up a bit.

**Comment by jenlampton** [comment]
@BWPanda can you copy/paste the conversation here for those who don't use gitter?

**Comment by BWPanda** [comment]
@jenlampton It seems you don't have to use Gitter to read it: https://gitter.im/backdrop/backdrop-issues?at=5ce3b7e813e9854e333eb5c0 (just accessed that from a private browser window where I wasn't logged in and it works fine). That's when I posted my original question, so scroll down from there.

**Comment by jenlampton** [comment]
I suppose I can do it for you:

```
How would I go about adding, or not adding, a word to the end of a translated string? E.g.:

$end = '';
if ($condition == TRUE) {
  $end = ' with a BANG';
}
t("This string ends@end.", array('@end' => $end))
That works, but: 1. the string inside t() isn't very pretty/readable, and 2. 'with a BANG' isn't translatable (and I don't think I can add a t() within a t(), right?


Justin Christoffersen @larsdesigns 02:28
@BWPanda I like to use sprintf() or vsprintf() depending on the number of strings.
sprintf(
%s, %s, 
t(’something @end’ , array(‘@end’ => $end)),
t(’something @else’ , array(‘@else’ => $else))
);

Justin Christoffersen @larsdesigns 02:34
I am not sure if that is to your point.

Olaf Grabienski @olafgrabienski 02:45
@klonos I've filed the issue for the missing project browser: #3803

Geoff St. Pierre @serundeputy 04:56
@BWPanda $end .= ' with a BANG'; the . is the concatenate operator and will glue the new string to the old.
https://www.php.net/manual/en/language.operators.string.php
```

**Comment by BWPanda** [comment]
> can you copy/paste the conversation here ~~for those who don't use gitter~~ so someone reading this issue can understand what's going on.

@jenlampton Oh, that's what you meant... Sorry. I've updated your comment with some better formatting for easier readability.

**Comment by klonos** [comment]
...and I have re-re-edited it, because people's comments should be formatted in quotes, and I have OCDs 😅

**Comment by BWPanda** [comment]
@klonos OCDs = Often Credited Dragons? 😉😆

**Comment by klonos** [comment]
I'm revitalising this issue with a fresh [PR](https://github.com/backdrop/backdrop/pull/2697) based on @klonos' work and notes above (hope you don't mind).

Don't mind it at all @BWPanda. I never came back to it, and in the meantime I have forgotten what plans/ideas I had for this to work "properly" 😅 👴 

> I think I've addressed all the issues you had with your initial attempt, but feedback welcome how to improve it more if necessary.

All issues but one:

![Screen Shot 2019-05-22 at 7 29 54 pm](https://user-images.githubusercontent.com/2423362/58163904-26563c80-7cc8-11e9-9125-df7cab1628eb.png)

^^ ...menu titles that include the word "menu" in their title, have the word "menu" italicised; while the ones that do not include it have the word "menu" in normal text. 🤔 

...the only way I can see this working consistently, is to strip the word "menu" from the title if found, and then re-add it (so that only the menu title is in italics - not the word "menu"). I know... nit-picking, but ...Often Credited Dragons 😜

**Comment by BWPanda** [comment]
Hmm, I thought based on your comments above that that's how you wanted it to work: menu titles italicised (and 'Administration menu' is the menu title)... Considering menu names can be edited, I think whatever people put as the title should be italicised, regardless whether that includes 'menu' or not. My 2c.

**Comment by klonos** [comment]
I think that after all this endless discussion, the best way (or compromise I should say) would be to not be appending the word "menu", but rather have it come before the menu title:

`A list of links for the menu %title`

...I realise that we will still have some duplication of the word "menu", but it won't be "_menu_ menu" ...so not as bad I guess.

This will be better translation-wise as well I think. Only a single string to translate rather than 2 slightly different versions of the same thing.

Thoughts?

**Comment by BWPanda** [comment]
Sure, that sounds easier. Will work on a new PR when I get a chance.

**Comment by jenlampton** [comment]
And add a colon.
The menu: menu isn't so bad.

**Comment by klonos** [comment]
...idea for a follow-up: perhaps as a generic solution, we should be looking into allowing to set "tags" for the entries that are listed in the "Add block" dialog. If that was possible, we could have added the tag `menu` to the menus, and make the search functionality take those into account. That way, we would not be needing to include the word "menu" in the visible description of the block, but the search thing (which was one of the original reasons for this issue here - the other one was to remove the word "hierarchical") would still work 😉

**Comment by BWPanda** [comment]
PR updated to be @klonos' much simpler version.

And I like the sound of tags, but definitely a separate issue.

**Comment by klonos** [comment]
RTBC! Less code++ 👍 

This looks way more consistent now (although slightly wordy); still way better than the previous situation. Thanks for working on this one @BWPanda 👍 

...and yes, certainly follow-up issue for adding optional tags to blocks.

**Comment by quicksketch** [comment]
I left a few comments on the PR: https://github.com/backdrop/backdrop/pull/2697/files#r289573756 if @BWPanda can approve those suggestions.

**Comment by BWPanda** [comment]
@quicksketch @klonos The reason I removed the t() function was because of this:

> You should never use t() to translate variables, such as calling `t($text);`...

https://api.backdropcms.org/api/backdrop/core%21includes%21bootstrap.inc/function/t/1

But if I'm understanding that wrong then happy for it to be added back in.

**Comment by quicksketch** [comment]
Thanks! Merged https://github.com/backdrop/backdrop/pull/2715 into 1.x and 1.13.x.
